include/revng/Support/GraphAlgorithms.h

```diff
@@ -196,3 +201,215 @@ auto exitless_scc_range(NodeTy Entry) {
 
   return make_filter_range(Range, Filter);
 }
+
+namespace detail {
+
+template<typename T>
+T stripInverseHelper(llvm::Inverse<T> &&);
+
+template<typename T>
+T stripInverseHelper(T &&);
+
+template<typename T>
+using StripInverse = decltype(stripInverseHelper(std::declval<T>()));
+
+// clang-format off
+template<typename T>
+concept SupportedGraph = std::is_pointer_v<T> ||
+                         (StrictSpecializationOf<T, llvm::Inverse> &&
+                          std::is_pointer_v<StripInverse<T>>);
+// clang-format on
+
+template<typename GraphType>
+using GT = llvm::GraphTraits<GraphType>;
+
+} // namespace detail
+
+template<detail::SupportedGraph RawGraphType>
+class GraphNodeOrRootHelper {
+public:
+  using GraphType = detail::StripInverse<RawGraphType>;
+
+  using NodeRef = typename detail::GT<RawGraphType>::NodeRef;
+  using ValueType = std::variant<GraphType, NodeRef>;
+  using EquivalentPair = std::pair<GraphType, NodeRef>;
+
+  using GraphTrait = detail::GT<RawGraphType>;
+  using NodeTrait = std::conditional_t<std::is_same_v<GraphType, RawGraphType>,
+                                       detail::GT<NodeRef>,
+                                       detail::GT<llvm::Inverse<NodeRef>>>;
+
+  using Iterator = VariedIterator<typename GraphTrait::nodes_iterator,
+                                  typename NodeTrait::ChildIteratorType>;
+
+public:
+  ValueType Value;
+
+public:
+  GraphNodeOrRootHelper(GraphType Graph) : Value(Graph) {}
+  GraphNodeOrRootHelper(llvm::Inverse<GraphType> Inv) : Value(Inv.Graph) {}
```

Comment:
`Inverse` is a decorator as there can be many others, you're not supposed to handle it explicitly.
If you need ad-hoc handling, we're probably doing it wrong.
See my other comment.

Reply:
Yep, but the problem itself is caused by the `llvm::Inverse` class - it has a constructor from the original graph type, but no way to convert back.

Here, since we have to get the fact of the iteration being "Inverse" inside `GraphNodeOrVirtualRoot` (the specialization trick:
```cpp
template<typename GraphType>
struct llvm::GraphTraits<llvm::Inverse<GraphNodeOrVirtualRoot<GraphType>> : llvm::GraphTraits<GraphNodeOrVirtualRoot<llvm::Inverse<GraphType>> {};
```
)
we run into the cases llvm's own algorithm try to convert these back-and-forth. And since only one conversion exists explicitly, the other one had to be provided, even if it pains me to do that.

Reply:
I'd like to hear @pfez on this.

Reply:
Me too :)

Reply:
Closing this, since I found a way to get rid of this without losing functionality.